### PR TITLE
ENH: attached rate and motif parameters to cogent3 tree object

### PR DESCRIPTION
**Pull request by YapengLang**
#27

### CODE REVIEW

**Comment by rmcar17** [comment]
@YapengLang could you add a test which checks for the parameters when `build_tree`/`fit_tree` are called?

**Comment by YapengLang** [comment]
> @YapengLang could you add a test which checks for the parameters when `build_tree`/`fit_tree` are called?

Sure, will do.

File: src/piqtree2/iqtree/_tree.py [hunk 1]
```diff
@@ -9,19 +9,74 @@
...
+
 
 def _rename_iq_tree(tree: cogent3.PhyloNode, names: Sequence[str]) -> None:
     for tip in tree.tips():
         tip.name = names[int(tip.name)]
 
 
-def _process_tree_yaml(tree_yaml: dict, names: Sequence[str]) -> cogent3.PhyloNode:
+def _intrude_edge_pars(tree: cogent3.PhyloNode, **kwargs: dict) -> None:
+    for node in tree.get_edge_vector():
```
Comment (GavinHuttley) [inline src/piqtree2/iqtree/_tree.py:30]: what's the function purpose, make it clearer that's about mapping edge parameters to cogent3's structure

File: src/piqtree2/iqtree/_tree.py [hunk 2]
```diff
@@ -9,19 +9,74 @@
...
+        # skip the rate parameters when node is the root
+        if node.is_root():
+            kwargs = {k: v for k, v in kwargs.items() if k == "mprobs"}
+            del node.params["edge_pars"]
+        node.params.update(kwargs)
+
+
+def _reform_edge_pars(tree: cogent3.PhyloNode, model: Model) -> cogent3.PhyloNode:
+    rate_pars = dict(
```
Comment (GavinHuttley) [inline src/piqtree2/iqtree/_tree.py]: as above, change name and add a comment describing what function is doing

File: tests/test_iqtree/test_fit_tree.py [hunk 3]
```diff
@@ -26,3 +26,21 @@ def test_fit_tree(three_otu: ArrayAlignment, iq_model: DnaModel, c3_model: str)
...
+    exclude = {"length", "ENS", "paralinear", "mprobs"}
+    expected_rates = {
+        k: v
+        for k, v in expected.tree.get_edge_vector()[0].params.items()
+        if k not in exclude
+    }
+    for k, v in expected_rates.items():
+        assert k in got2.get_edge_vector()[0].params
+        assert got2.get_edge_vector()[0].params[k] == pytest.approx(
+            v, abs=1e-1,
```
Comment (GavinHuttley) [inline tests/test_iqtree/test_fit_tree.py]: i think this value too permissive, can you remove and let me see the failures

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
nice one! few changes, clearly need to understand that numerical issue

File: tests/test_iqtree/test_fit_tree.py (hunk 3 above)
Comment (YapengLang) [reply tests/test_iqtree/test_fit_tree.py]: > i think this value too permissive, can you remove and let me see the failures

Hi @GavinHuttley , I've removed the tolerance setting and brought out the failures. Each fit of K80, GTR, TN93 and HKY85 showed one rate parameter that did not match the expected.

File: src/piqtree2/iqtree/_tree.py [hunk 4]
```diff
@@ -9,19 +9,77 @@
...
+    # cogent3.PhyloNode
+    for node in tree.get_edge_vector():
+        # skip the rate parameters when node is the root
+        if node.is_root():
+            kwargs = {k: v for k, v in kwargs.items() if k == "mprobs"}
+            del node.params["edge_pars"]
+        node.params.update(kwargs)
+
+
+def _format_edge_pars(tree: cogent3.PhyloNode, model: Model) -> None:
```
Comment (GavinHuttley) [inline src/piqtree2/iqtree/_tree.py]: This is not just a format operation, perhaps `_edge_pars_for_cogent3()`?

File: src/piqtree2/iqtree/_tree.py [hunk 5]
```diff
@@ -9,19 +9,77 @@
...
+    # converts the strings of rate and motif parameters into a dictionary format
+    rate_pars = dict(
+        zip(RATE_PARS, map(float, tree.params["edge_pars"]["rates"].split(", "))),
+    )
+    motif_pars = {
```
Comment (GavinHuttley) [inline src/piqtree2/iqtree/_tree.py]: I think all type casting to float / int should be done within `_process_tree_yaml()`.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
Thanks for the revisions! The test failures due to the float comparisons are all ~1e-6. If you set the tolerance to 1e-5 do they all pass? That level of tolerance for a parameter estimate on limited data seems reasonable.

**Comment by YapengLang** [comment]
> Thanks for the revisions! The test failures due to the float comparisons are all ~1e-6. If you set the tolerance to 1e-5 do they all pass? That level of tolerance for a parameter estimate on limited data seems reasonable.

hi @GavinHuttley , setting `abs = 1e-5` or `1e-6` does not solve. The largest difference arises from the GTR fit, where one rate estimate is 3.9724, but the expected value is around 3.9915.

Alternatively, I can try setting `rel = 1e-2`. This would allow the GTR fit to pass with `3.9724 == 3.9915 ± 4.0e-02`, and the other fits would also pass (please see commit 700f39869b3550f638a59e1d9cf9293244ea752b).

**Comment by GavinHuttley** [comment]
Thanks for clarifying @YapengLang !

**Review by GavinHuttley — APPROVED** [review]
Looks good!

**Review by rmcar17 — APPROVED** [review]

**Comment by rmcar17** [comment]
Thanks @YapengLang, it will auto-merge once the tests all pass!

**Pull request merged**